kernel/sync_preemption/src/lib.rs

```diff
@@ -0,0 +1,23 @@
+#![feature(negative_impls)]
+#![no_std]
+
+pub type Mutex<T> = sync::Mutex<DisablePreemption, T>;
+pub type MutexGuard<'a, T> = sync::MutexGuard<'a, DisablePreemption, T>;
+
+/// A deadlock prevention method that disables preemption.
+#[derive(Copy, Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct DisablePreemption {}
+
+impl !Send for DisablePreemption {}
+
+impl sync::DeadlockPrevention for DisablePreemption {
+    #[inline]
+    fn enter() {
+        preemption::disable_preemption()
```

Comment:
as i mentioned in the comments on the `DeadlockPrevention` trait, this isn't correct, and unfortunately it cannot be correct due to a lacking interface. The issue here is that preemption can't just be arbitrarily enabled and disabled; to do it safely we need to maintain a count and ensure that it's only enabled or disabled when that count changes to or from 0.

That's the reason I introduced a `PreemptionGuard` type, and also why there is no other direct way to manually enable/disable preemption -- the guard type semantic is useless if there is also a broken way to circumvent it.

This is similar to what you said in your "FIXME" comment in the `DisableIrq` implementation; we should modify the `DeadlockPrevention` interface to allow guard types and other stateful contexts to be propagated across `enter` and `exit` calls.